### PR TITLE
add asqa benchmark

### PR DESCRIPTION
**Pull request by bugtig6351**
https://github.com/gomate-community/rageval/issues/63#issuecomment-2003796017

### CODE REVIEW

File: benchmarks/ASQA/evaluate.py
Comment (faneshion) [inline benchmarks/ASQA/evaluate.py:1]: Maybe we should implement the benchmark with Class, as with ELI5.

**Review by faneshion — CHANGES_REQUESTED** [review]

File: benchmarks/ASQA/evaluate.py
Comment (bugtig6351) [reply benchmarks/ASQA/evaluate.py:1]: I think that a benchmark should focus solely on the evaluation, while the generation method should be implemented elsewhere. So I have designed an abstract class as follows: 

```python
class BaseBenchmark(ABC):

    name: str
    metrics: List[str]

    def __init__() -> None:
        ...

    @abstractmethod
    def load_dataset() -> Dataset:
        ...

    @abstractmethod
    def evaluate() -> Dataset:
        ...
    
    @abstractmethod
    def save_result() -> None:
        ...

    def get_metric(name: str) -> Callable:
        module = importlib.import_module(f"rageval.metrics")
        return getattr(module, name)
```

`load_dataset()` is used to  receive the data path, check the dataset format and load it. `evaluate()` is used to test the metric performance and attach the results to the dataset, and `save_result()` saves the result files. And we could consider adding a function in the `utils.py` to display the evaluation results by inputing a dataset with socres and outputing a result table.

And there is a problem that different metrics may require different ground truths. For instance, in ASQA benchmark, the `em` metric compares `short_answers` and `answers`, while `rouge` uses `long_answers` as ground truths. How do we deal with it? Now I am resolving this by just rebuilding dataset[`gt_answers` ] in `evaluate` function when calculating different metrics, but there is only one `gt_answers` can be saved in single result file.

File: benchmarks/ASQA/generate.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        find = re.compile(pattern).search(pred)
+        if find:
+            pred = find.group(1)
+            break
+    if find is None:
+        logging.warning(f"Cannot extract key information from the response: {pred}")
+    pred = re.sub(r'\(\d+\)\s', '', pred) # remove the index numbers
+    return pred
+
+def generete_answers(engine: InstructGPT, dataset: Dataset) -> Dataset:
```
Comment (faneshion) [inline benchmarks/ASQA/generate.py:103]: `generate_answers`

File: benchmarks/ASQA/generate.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    dataset = dataset['dev'].select(range(args.max_num_examples))
+
+    print("Init ASQA dataset...")
+    os.environ['OPENAI_API_KEY'] = args.api_key
+    engine = InstructGPT(args.model, 
+                         _api_key_env_var = 'OPENAI_API_KEY', 
+                         max_tokens=args.max_new_tokens)
+
+    print("Start generate answers...")
+    dataset = generete_answers(engine, dataset)
```
Comment (faneshion) [inline benchmarks/ASQA/generate.py:136]: `generate_answers`

File: benchmarks/ASQA/README.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+`--max_num_examples`: The maximum number of examples used in generate answers.
+
+`--max_new_tokens`: The maximum number of tokens that can be generated for answering questions.
+
+`--output_dir`: Directory where the generated answers and the final results will be saved.
+
+`--model`: The OpenAI GPT model to use, e.g., gpt-3.5-turbo-instruct.
+
+`--api_key`: Your OpenAI API key.
+
```
Comment (faneshion) [inline benchmarks/ASQA/README.md:21]: add performances in this file:

### Performances:
|  methods   | metric-1  | metric-2 |
|  ----  | ----  | ----  |
| XXXX  | 单元格 |单元格 |
| XXXX  | 单元格 |单元格 |

**Review by faneshion — CHANGES_REQUESTED** [review]

File: benchmarks/ASQA/README.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
+# ASQA BENCHMARK
+
+This benchmark is designed to evaluate the performance of the [ASQA dataset](https://huggingface.co/datasets/din0s/asqa). For `generate.py`, we followed [FLARE](https://github.com/jzbjyb/FLARE), using `gpt-3.5-turbo-instruct` with no retrieval settings.
+
+## Usage
+
+1. Prepare your model output results in the `output` directory. You can just attatch your `answers` to the end of each example in the origin ASQA dataset as what we did in the `dataset.json` file.
+2. Run `run.sh` to start dataset generation. The result will be saved in `output` directory, named `results.json`. And the `result_datasets.json` file is the json dump of the detailed result.
+
```
Comment (Wenshansilvia) [inline benchmarks/ASQA/README.md:9]: 1. Upload dataset.json file to HuggingFace and provide a link here. 
2. Maybe its better to seperate dataset generation step and evaluate step.

File: benchmarks/ASQA/output/dataset.jsonl
Comment (Wenshansilvia) [inline benchmarks/ASQA/output/dataset.jsonl:1]: Large files should be upload to Huggingface. And functions that take these files as input should be able to read from local file path as well as huggingface dataset name(load_dataset("name")).

**Review by Wenshansilvia — CHANGES_REQUESTED** [review]

**Review by faneshion — APPROVED** [review]

**Review by Wenshansilvia — APPROVED** [review]
